ng-dev/release/notes/changelog.ts

```diff
@@ -0,0 +1,127 @@
+import {existsSync, readFileSync, writeFileSync} from 'fs';
+import {join} from 'path';
+import * as semver from 'semver';
+import {GitClient} from '../../utils/git/git-client';
+
+/** Project-relative path for the changelog file. */
+export const changelogPath = 'CHANGELOG.md';
+
+/** Project-relative path for the changelog archive file. */
+export const changelogArchivePath = 'CHANGELOG_ARCHIVE.md';
+
+/** A marker used to split a CHANGELOG.md file into individual entries. */
+export const splitMarker = '<!-- CHANGELOG SPLIT MARKER -->';
+
+/**
+ * A string to use between each changelog entry when joining them together.
+ *
+ * Since all every changelog entry's content is trimmed, when joining back together, two new lines
+ * must be placed around the splitMarker to create a one line buffer around the comment in the
+ * markdown.
+ * i.e.
+ * <changelog entry content>
+ *
+ * <!-- CHANGELOG SPLIT MARKER -->
+ *
+ * <changelog entry content>
+ */
+const joinMarker = `\n\n${splitMarker}\n\n`;
+
+/** A RegExp matcher to extract the version of a changelog entry from the entry content. */
+const versionAnchorMatcher = new RegExp(`<a name="(.*)"></a>`);
+
+/** An individual changelog entry. */
+interface ChangelogEntry {
+  content: string;
+  version: semver.SemVer;
+}
+
+export class Changelog {
+  /** The absolute path to the changelog file. */
+  readonly changelogPath = join(this.git.baseDir, changelogPath);
+  /** The absolute path to the changelog archive file. */
+  readonly changelogArchivePath = join(this.git.baseDir, changelogArchivePath);
+
+  /** The changelog entries in the CHANGELOG.md file. */
+  private entries = this.getEntriesFor(this.changelogPath);
+  /** The changelog entries in the CHANGELOG_ARCHIVE.md file. */
+  private archiveEntries = this.getEntriesFor(this.changelogArchivePath);
+  /** Whether the changelog archive has entries upon intialization. */
+  private alreadyHasChangelogArchive = this.archiveEntries.length !== 0;
+
+  constructor(private git: GitClient) {}
+
+  /** Prepend a changelog entry to the changelog. */
+  prependEntryToChangelog(entry: string) {
+    this.entries.unshift(parseChangelogEntry(entry));
+    this.writeToChangelogFiles();
+  }
+
+  /**
+   * Move all changelog entries from the CHANGELOG.md file for versions prior to the provided
+   * version to the changelog archive.
+   */
+  moveEntriesPriorToVersionToArchive(version: semver.SemVer) {
```

Comment:
Can you elaborate on why this concept is used, compared to just limiting the entries to a certain amount (like you did with the changelog github action). That way we can limit entries in the changelog to a fixed number and not have to worry about version-specific details.

Reply:
I can put this in a comment, but it made the most sense to me for us to migrate all of the entries from versions after something like the LTS period ends, or upon its entry into the LTS version.

For dev-infra it made sense to do it based on the number of entries because we have roughly one entry per week.  The entries in our normal changelogs are representative of actual versioning, so the context for whether something matters in the changelog is actually the version rather than the date it was created.

Reply:
That's fair 👍 this comment on the PR is sufficient I assume